File: 41 First Missing Positive/41 LL.cpp
#include <vector>
#include <iostream>
using namespace std;

/*
https://leetcode.com/problems/first-missing-positive/

Given an unsorted integer array, find the first missing positive integer.

For example,
Given [1,2,0] return 3,
and [3,4,-1,1] return 2.

Your algorithm should run in O(n) time and uses constant space.
*/

class Solution
{
public:
	int firstMissingPositive(vector<int>& nums)
	{
		for (int i = 0; i < nums.size(); ++i)
		{
			while ((0 < nums[i]) && (nums[i] != (i + 1)) && (nums[i] <= nums.size()))
			{
				int cache = nums[nums[i] - 1];
				nums[nums[i] - 1] = nums[i];
				if (cache == nums[i]) break;	// prevent infinity loop
				nums[i] = cache;
			}
		}
		int reslut = 0;
		for (; reslut < nums.size(); ++reslut)
		{
			if (nums[reslut] != (reslut + 1))
			{
				break;
			}
		}
		return reslut + 1;
	}
};

/*
idea: put number to corresponding order position
complexity: Time O(N * target)
*/

int main(void)
{
	vector<int> nums;
	nums.push_back(1);
	nums.push_back(1);
	Solution engine;
	cout << engine.firstMissingPositive(nums) << '\n';
	return 0;
}

File: 41 First Missing Positive/41 LL.py
class Solution(object):
	def firstMissingPositive(self, nums):
		for i in xrange(len(nums)):
			pos = nums[i] - 1
			while pos != i and pos >= 0 and pos < len(nums) and nums[i] != nums[pos]:
				nums[i], nums[pos] = nums[pos], nums[i]
				pos = nums[i] - 1
		for i in xrange(len(nums)):
			if nums[i] != (i + 1):
				return (i + 1)
		return len(nums) + 1

engine = Solution()
print engine.firstMissingPositive([2, 1])